source/Halibut/Transport/Streams/WriteIntoMemoryBufferStream.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Halibut.Transport.Observability;
+
+namespace Halibut.Transport.Streams
+{
+    public class WriteIntoMemoryBufferStream : Stream
+    {
+        readonly MemoryStream memoryStream;
+        readonly Stream sinkStream;
+        readonly long writeIntoMemoryLimitBytes;
+        readonly OnDispose onDispose;
+        bool usingSinkStream;
+
+        public WriteIntoMemoryBufferStream(Stream sinkStream, long writeIntoMemoryLimitBytes, OnDispose onDispose)
+        {
+            memoryStream = new MemoryStream();
+            this.sinkStream = sinkStream;
+            this.writeIntoMemoryLimitBytes = writeIntoMemoryLimitBytes;
+            this.onDispose = onDispose;
+        }
+        
+        public long BytesWrittenIntoMemory => memoryStream.Length;
+
+        public override bool CanRead => false;
+        public override bool CanWrite => sinkStream.CanWrite;
+        public override bool CanSeek => false;
+        public override bool CanTimeout => false;
+
+        public override long Length => throw new NotSupportedException();
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
```

Comment:
May as well write the rest of the cache if any to the underlying stream here as well, so no one can get it wrong.

Reply:
Sure. I wrote a test for this as well 😁